### DISCUSSION TITLE
[PSR-12] space after closing parenthesis for alternative control structure syntax

### DISCUSSION THREAD

**Discussion by herndlm** [squizlabs/PHP_CodeSniffer#3391] [Q&A] [answered]
Hi, quick question / potentially longer discussion ;)

This is about a potential sniff [PSR-12](https://www.php-fig.org/psr/psr-12/) inconsistency with [alternative syntax for control structures](https://www.php.net/manual/en/control-structures.alternative-syntax.php).

PSR-12 states (under 5. Control Structures)
> There MUST be one space between the closing parenthesis and the opening brace

And apparently `Squiz.ControlStructures.ControlSignature` is enforcing that. I noticed that it complains for the following
```php
<?php if ($foo)): ?>
   <?= esc_html($foo) ?>
<?php endif; ?>
```
with
> Squiz.ControlStructures.ControlSignature.SpaceAfterCloseParenthesis: Expected 1 space(s) after closing parenthesis; found 0

So it wants that space after the closing paranthesis and _before the colon_:
```php
<?php if ($foo)) : ?>
   <?= esc_html($foo) ?>
<?php endif; ?>
```

But I was wondering - is that really on purpose or a potential bug? At least PSR-12 seems to be not specifing that as it deals only with spaces before the opening brace, right?

Btw. unrelated: thank you everyone for the hard work on phpcs!

**Comment by gsherwood** [comment, marked as answer]
First, I strongly recommend not trying to enforce PSR12 on template code. Alternate syntax rules were excluded from the PSR standards on purpose so they wouldn't mix template standards with other coding standards. Hence the use of "brace" everywhere.

But if you want to use it, PHPCS sees the opening brace of this alternate IF syntax as the `:` and the closing brace as the `endif`. The sniffs that check control structures don't care if you're using regular or alternate syntax - they just check that the opening and closing scope tokens are in the right spots. This behaviour is completely intentional throughout PHPCS.

You may also want to consider excluding this specific sniff or error code from checking your template files if you still want the rest of PSR-12 applied. You'd need a `phpcs.xml` file containing something like:
```xml
 <rule ref="Squiz.ControlStructures.ControlSignature.SpaceAfterCloseParenthesis">
    <exclude-pattern>*.tpl</exclude-pattern>
 </rule>
```
More info here: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-Ruleset

**Reply by herndlm** [reply]
Makes sense, thank you for the detailed explanation!
